### DISCUSSION TITLE
Difference between warmStart and mip_start in CBC?

### DISCUSSION THREAD

**Discussion by ebongo1** [coin-or/pulp#564] [Q&A]
I am trying to utilize the warmStart functionality for CBC using pulp, but noticed that the results don't look how I'd expect. I would expect to see the run time to be fast using warmStart and possibly a better solution. However, I am seeing long run times and the same solutions, or even worse ones sometimes. I did a bit of research and noticed there is a difference between the warmStart and mip_start feature in CBC: Reoptimization (or warmstarting): the solution of a problem based on information available from solving a similar problem
Optimization via a MIP start: the solution of a problem based on providing an initial feasible solution.

However, when I looked in the coin_api.py python file located in pulp/pulp/apis, I noticed this comment: ":param bool mip_start: deprecated for warmStart". Does this mean that pulp is only using the warmStart feature and not the mip_start? Am I misunderstanding how these two work? Are they intended to work the same way?

**Comment by pchtsp** [comment]
When using the warmStart parameter in pulp, we call the mip_start option
for CBC. I wasn't aware there was also a warmStart option on CBC.

For us, warmStart implies providing a valid solution. Although some solvers
(like CPLEX and GUROBI) accept partial /incomplete solutions. I don't think
CBC does this, at least with the mip_start option.

Can you share the CBC docs where they mention these two configurations?

Thanks!

F.

**Reply by ebongo1** [reply]
In this link it shows that the COIN_CMD solver https://coin-or.github.io/pulp/technical/solvers.html has a warmstart and mip_start argument, however I can't find specific CBC docs for this. But it sounds like you are just treating warmstart the same as mip_start?

It just seemed a bit confusing since there is a difference between warm starting and a mip start. A warm start includes:
-	Providing a feasible, but not optimal solution
-	Providing a partial solution (partially feasible) and
-	Providing an invalid solution.

In contrast, a MIP start only includes the first bullet; providing a feasible solution. So, all MIP starts are warm starts but not all warm starts are MIP starts. In branch and bound, providing the initial root node (LP relaxation) solve is like a warm start and the initial integer solution (farther down in the tree) is like a MIP start. 


Also, I noticed that the argument "keepFiles" changes the way the warm start works (See attached log files, one where this arg is set to True and one where it's set to False). When it is set to False, the log says "will open mipstart file .\C:\Users\ebongo\AppData\Local\Temp\71619a5f622e494190391f963a562e6c-pulp.mst.", but then it doesn't log anything else about warmstarting.

When it's set to True, it says the following: 
"will open mipstart file .\ACE_Optimization-pulp.mst.
mipstart values read for 14665 variables.
.
.
.
Cbc0045I Warning: mipstart values could not be used to build a solution.
Cbc0045I Trying just fixing integer variables.
Cbc0045I Still no good."


Either way, the warmstart seems to be working a bit weird. Can you please clarify why this "keepFiles" arg would alter the way warmStart is working? And why in general the warmStart seems to not be working? 

[solver _warmstart_keepFilesFalse.txt](https://github.com/coin-or/pulp/files/9189998/solver._warmstart_keepFilesFalse.txt)
[solver_warmstart_keepFilesTrue.txt](https://github.com/coin-or/pulp/files/9189999/solver_warmstart_keepFilesTrue.txt)

**Reply by pchtsp** [reply]
Hello again!
`mip_start` argument is only there because it's deprecated (it says so in those same docs). I guess we can just take it out in the next release and it will avoid confusion. About the actual definitions of `warm_start` and `mip_start`, I'm not super sure there is one. I know that CPLEX and GUROBI accept infeasible or partially complete solutions and will try to complete them (this would be `warm_start` according to your definition). CBC I think is a little more sensitive to the quality of the solution (although from your log messages, they also try to complete it). In any case, I think it's good that we call the standard argument `warm_start` and then it's up to the solver to determine how it will use the provided solution and if it's compatible.

With regards to the `keepFiles` behavior, I recommend you create an issue in the [cbc repo](https://github.com/coin-or/cbc/issues), since there appears to be an problem with the way they get use the `mst` file paths. I see in your log files they're using `.\C:\Users\ebongo\AppData\Local\Temp\71619a5f622e494190391f963a562e6c-pulp.mst` when they should be using `C:\Users\ebongo\AppData\Local\Temp\71619a5f622e494190391f963a562e6c-pulp.mst` (i.e., without the first two characters). This could be a cause for the issue because then cbc would be unable to find the file (this doesn't happen when the path is relative).
